### PR TITLE
refactor(parser): introduce a compactible layer for parser v2 to call v1

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

* Introduce `ParseV1` as a compatible trait for parser v2
* Write `expr_case` as an example

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by BugenZhao — COMMENTED** [review]
👀 May I know the relationship versus #17019?

**Comment by TennyZhuang** [comment]
> 👀 May I know the relationship versus #17019?

The PR is based on that.

#17019 replace the underlying input stream of parser v2 from `&[TokenWithLocation]` to `parser_v1::Parser<'a>`.

The PR introduce a compatible trait, to allow `S: TokenStream` export underlying `parser_v1::Parser`, and call methods on it.

Based on the PR, we can refactor the recursive terms in the parser. E.g., we create a fake v2 `expr` which calls v1 internally, then we can write a real v2 `expr_case` based on the fake `expr`.

File: src/sqlparser/src/parser_v2/expr.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    })
+    .parse_next(input)
+}
+
+pub fn expr_case<S>(input: &mut S) -> PResult<Expr>
+where
+    S: TokenStream,
+{
+    let parse = (
+        opt(expr),
```
Comment (wangrunji0408) [inline src/sqlparser/src/parser_v2/expr.rs:35]: In terms of semantics, this function should start parsing from `CASE` and the caller should not consume `CASE` before calling it.

```suggestion
pub fn expr_case<S>(input: &mut S) -> PResult<Expr>
where
    S: TokenStream,
{
    let parse = (
        Keyword::CASE,
        opt(expr),
```

**Review by wangrunji0408 — COMMENTED** [review]
Rest LGTM!

File: src/sqlparser/src/parser_v2/expr.rs (hunk 1 above)
Comment (TennyZhuang) [reply src/sqlparser/src/parser_v2/expr.rs:35]: Unfortunately we have to following the v1's behavior.

Another problem is `dispatch!` is much more efficient than `alt`, even in v2, we have to use the style.

Comment (TennyZhuang) [reply src/sqlparser/src/parser_v2/expr.rs:35]: The only thing I can do is rename it to `expr_case_suffix`.

Comment (wangrunji0408) [reply src/sqlparser/src/parser_v2/expr.rs:35]: Okay let's correct this later.

`dispatch!` can also be used with `peek(keyword)` which don't consume the token. 😄

**Review by wangrunji0408 — APPROVED** [review]

**Pull request merged**